### PR TITLE
feat(tool): add sql database query tool

### PR DESCRIPTION
**Pull request by abughali**
### Which issue(s) does this pull-request address?

New SQL Database Tool.

### Description

This pull request introduces a new SQL tool to the framework. The following updates are included:

- Main tool code under src/tools/database
- Updated environment template for multiple database dialects (SQLite, MySQL, PostgreSQL, etc.)
- Integrated the sequelize and sqlite3 package in dev and peer dep.
- Unit and E2E testing.
- Sample SQL Agent.
- Documentation: updated main README.md. Added docs/sql-tool.md.

### Checklist

- [x] I have read the [contributor guide](https://github.com/i-am-bee/bee-agent-framework/blob/main/CONTRIBUTING.md)
- [x] Linting passes: `yarn lint` or `yarn lint:fix`
- [x] Formatting is applied: `yarn format` or `yarn format:fix`
- [x] Unit tests pass: `yarn test:unit`
- [x] E2E tests pass: `yarn test:e2e`
- [x] Tests are included 
- [x] Documentation is changed or added
- [x] Commit messages and PR title follow [conventional commits](https://www.conventionalcommits.org/en/v1.0.0/#summary)

### CODE REVIEW

File: src/tools/database/sql.ts [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+  ToolInput,
+  BaseToolOptions,
+  BaseToolRunOptions,
+  JSONToolOutput,
+} from "@/tools/base.js";
+import { z } from "zod";
+import { getMetadata } from "@/tools/database/metadata.js";
+import { isSupported, connectSql } from "@/tools/database/connection.js";
+
+type ToolOptions = { provider: string } & BaseToolOptions;
```
Comment (Tomas2D) [inline src/tools/database/sql.ts:28]: Use strong typing for the provider.

File: src/tools/database/sql.ts [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+  }
+
+  protected async _run(
+    { query }: ToolInput<this>,
+    _options?: ToolRunOptions,
+  ): Promise<JSONToolOutput<any>> {
+    const provider = this.options.provider;
+
+    if (!isSupported(provider)) {
+      throw new Error(`Unsupported database provider: ${provider}`);
```
Comment (Tomas2D) [inline src/tools/database/sql.ts:54]: always throw an instance of the `ToolError` class or a class that extends it

Comment (Tomas2D) [reply src/tools/database/sql.ts:54]: also you should do this check in the constructor to throw as soon as possible

File: src/tools/database/sql.ts [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+          success: false,
+          message: `No rows selected`,
+        });
+      }
+    } catch (error) {
+      const errorMessage = `Based on this database schema structure: ${metadata}, 
+      generate a correct query that retrieves data using the appropriate ${provider} dialect. 
+      The original request was: ${query}, and the error was: ${error.message}.`;
+
+      return new JSONToolOutput({ error: errorMessage });
```
Comment (Tomas2D) [inline src/tools/database/sql.ts:86]: You should throw `throw new ToolError(errorMessage);`

File: src/tools/database/connection.ts [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+      port: dbPort,
+      schema: dbSchema,
+      storage: dbStorage,
+      logging: false,
+    });
+
+    await sequelize.authenticate();
+    return sequelize;
+  } catch (error) {
+    throw Error(`Unable to connect to the SQL database: ${error.message}`);
```
Comment (Tomas2D) [inline src/tools/database/connection.ts:45]: `ToolError` with `isRetryable = false`

File: src/tools/database/connection.ts [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { Sequelize, Dialect } from "sequelize";
+
+let dbSchema: string | undefined = undefined;
+
+export async function connectSql(): Promise<Sequelize> {
```
Comment (Tomas2D) [inline src/tools/database/connection.ts:21]: This should be part of the class as a cached property.

File: package.json [hunk 6]
```diff
@@ -115,7 +115,9 @@
     "groq-sdk": "^0.7.0",
     "ollama": "^0.5.8",
     "openai": "^4.56.0",
-    "openai-chat-tokens": "^0.2.8"
+    "openai-chat-tokens": "^0.2.8",
+    "sequelize": "^6.37.3",
+    "sqlite3": "^5.1.7"
```
Comment (Tomas2D) [inline package.json:120]: `SQLite` should not be there; it is up to the user to pick one.

**Review by Tomas2D — CHANGES_REQUESTED** [review]

File: src/tools/database/sql.ts (hunk 2 above)
Comment (abughali) [reply src/tools/database/sql.ts]: not needed after strong typing of provider

File: src/tools/database/metadata.ts [hunk 7]
```diff
@@ -0,0 +1,135 @@
...
+          SELECT t.table_name AS "tableName", c.column_name AS "columnName", 
+                 c.data_type AS "dataType"
+          FROM all_tables t
+          JOIN all_tab_columns c ON t.table_name = c.table_name
+          WHERE t.owner = upper('${schemaName}');
+        `;
+
+    default:
+      throw new ToolError(`Provider ${provider} is not supported`, [], {
+        isRetryable: false,
```
Comment (Tomas2D) [inline src/tools/database/metadata.ts:132]: add `isFatal: true`

File: src/tools/database/sql.ts [hunk 8]
```diff
@@ -0,0 +1,116 @@
...
+    this.register();
+  }
+
+  private _sequelize: Sequelize | null = null;
+
+  @Cache()
+  protected async connection(): Promise<Sequelize> {
+    this._sequelize = await connectSql();
+    return this._sequelize;
+  }
```
Comment (Tomas2D) [inline src/tools/database/sql.ts:58]: Rebase against the `main` branch and do the following.

```typescript
  @Cache()
  protected get connection(): Promise<Sequelize> {
     return await connectSql();
  }
```

File: src/tools/database/sql.ts [hunk 9]
```diff
@@ -0,0 +1,116 @@
...
+  public async destroy(): Promise<void> {
+    if (this._sequelize) {
+      try {
+        await this._sequelize.close();
+        this._sequelize = null;
+      } catch (error: any) {
+        throw new ToolError(`Failed to close the database connection: ${error.message}`);
+      }
+    }
+  }
```
Comment (Tomas2D) [inline src/tools/database/sql.ts:115]: Rebase against the `main` branch and do the following.

```typescript
public async destroy() {
   // @ts-expect-error protected property
   const cache = Cache.getInstance(this, "connection")
   const entry = cache.get()

   if (entry) {
      cache.clear()
      
      try {
        await entry.data.close();
      } catch (error) {
        throw new ToolError(`Failed to close the database connection`, [error]);
      }
   }
}
```

**Review by Tomas2D — CHANGES_REQUESTED** [review]

File: src/tools/database/sql.ts (hunk 9 above)
Comment (abughali) [reply src/tools/database/sql.ts]: this doesn't close the connection when process.exit(0) follows

Comment (Tomas2D) [reply src/tools/database/sql.ts]: There might be a typo in the code. But it does the same thing as your previous code.

**Review by Tomas2D — CHANGES_REQUESTED** [review]
Allow a user to pass connection config instead of loading them from ENV.

File: src/tools/database/sql.ts [hunk 10]
```diff
@@ -0,0 +1,134 @@
...
+    try {
+      const sequelize = new Sequelize(connection);
+
+      await sequelize.authenticate();
+      return sequelize;
+    } catch (error) {
+      throw new ToolError(`Unable to connect to database: ${error}`, [], {
+        isRetryable: false,
+      });
+    }
```
Comment (Tomas2D) [inline src/tools/database/sql.ts:116]: move the content of this method directly to `connection`

Comment (Tomas2D) [reply src/tools/database/sql.ts:116]:
```
throw new ToolError(`Unable to connect to database`, [error], {
        isRetryable: false,
        isFatal: true,
      });
```

File: src/tools/database/sql.ts [hunk 11]
```diff
@@ -0,0 +1,134 @@
...
+  BaseToolOptions,
+  BaseToolRunOptions,
+  JSONToolOutput,
+} from "@/tools/base.js";
+import { z } from "zod";
+import { Sequelize, Options } from "sequelize";
+import { getMetadata } from "@/tools/database/metadata.js";
+import { Cache } from "@/cache/decoratorCache.js";
+
+type Provider = "mysql" | "mariadb" | "postgres" | "mssql" | "db2" | "sqlite" | "oracle";
```
Comment (Tomas2D) [inline src/tools/database/sql.ts:30]: I would move this `Provider` type to `metadata.ts` and export it from there.

File: src/tools/database/metadata.ts [hunk 12]
```diff
@@ -0,0 +1,142 @@
...
+
+    default:
+      throw new ToolError(`Provider ${provider} is not supported`, [], {
+        isFatal: true,
+        isRetryable: false,
+      });
+  }
+}
+
+function getDefaultSchema(provider: string): string {
```
Comment (Tomas2D) [inline src/tools/database/metadata.ts:128]: `function getDefaultSchema(provider: Provider): string {`

File: src/tools/database/metadata.ts [hunk 13]
```diff
@@ -0,0 +1,142 @@
...
+      .map((columns) => columns.join(", "))
+      .join("; ");
+  } catch (error) {
+    throw new ToolError(`Error initializing metadata: ${error}`, [], {
+      isRetryable: false,
+    });
+  }
+}
+
+function getMetadataQuery(provider: string, schema?: string): string {
```
Comment (Tomas2D) [inline src/tools/database/metadata.ts:56]: `function getMetadataQuery(provider: Provider, schema?: string): string {`

File: src/tools/database/metadata.ts [hunk 14]
```diff
@@ -0,0 +1,142 @@
...
+
+export interface Metadata {
+  tableName: string;
+  columnName: string;
+  dataType: string;
+}
+
+export async function getMetadata(
+  sequelize: Sequelize,
+  provider: string,
```
Comment (Tomas2D) [inline src/tools/database/metadata.ts:28]: `provider: Provider,`

File: src/tools/database/metadata.ts [hunk 15]
```diff
@@ -0,0 +1,142 @@
...
+function getDefaultSchema(provider: string): string {
+  switch (provider) {
+    case "postgres":
+      return "public";
+    case "mssql":
+      return "dbo";
+    case "db2":
+    case "oracle":
+      throw new ToolError(`Schema name is required for ${provider}`, [], {
+        isRetryable: false,
```
Comment (Tomas2D) [inline src/tools/database/metadata.ts:137]: add `isFatal: true`

File: src/tools/database/sql.ts [hunk 16]
```diff
@@ -0,0 +1,134 @@
...
+  inputSchema() {
+    return z.object({
+      query: z.string({ description: "The SQL query to be executed." }).min(1),
+    });
+  }
+
+  static {
+    this.register();
+  }
+
```
Comment (Tomas2D) [inline src/tools/database/sql.ts:53]: Add a constructor and verify if the given schema is `supported.` It is better to throw an error as soon as possible.

**Review by Tomas2D — CHANGES_REQUESTED** [review]

**Review by Tomas2D — APPROVED** [review]
Great work 🚀 Thank you for the contribution 👍🏻

**Pull request merged**